workmail-message-flow-state-machine/src/app.py

```diff
@@ -0,0 +1,168 @@
+            },
+            "subject" : "Hello From Amazon WorkMail!",              # String containing email subject (Truncated to first 256 chars)"
+            "messageId": "00000000-0000-0000-0000-000000000000",    # String containing message id for retrieval using workmail flow API
+            "invocationId": "00000000000000000000000000000000",     # String containing the id of this lambda invocation. Useful for detecting retries and avoiding duplication
+            "flowDirection": "INBOUND",                             # String indicating direction of email flow. Value is either "INBOUND" or "OUTBOUND"
+            "truncated": false                                      # boolean indicating if any field in message was truncated due to size limitations
+        }
+
+    context: object, required
+    Lambda Context runtime methods and attributes
+
+    Attributes
+    ----------
+
+    context.aws_request_id: str
+         Lambda request ID
+    context.client_context: object
+         Additional context when invoked through AWS Mobile SDK
+    context.function_name: str
+         Lambda function name
+    context.function_version: str
+         Function version identifier
+    context.get_remaining_time_in_millis: function
+         Time in milliseconds before function times out
+    context.identity:
+         Cognito identity provider context when invoked through AWS Mobile SDK
+    context.invoked_function_arn: str
+         Function ARN
+    context.log_group_name: str
+         Cloudwatch Log group name
+    context.log_stream_name: str
+         Cloudwatch Log stream name
+    context.memory_limit_in_mb: int
+        Function memory
+
+        https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+
+    Returns
+    -------
+    Amazon WorkMail Sync Lambda Response Format
+    For more information, see https://docs.aws.amazon.com/workmail/latest/adminguide/lambda.html#synchronous-schema
+        return {
+          'actions': [                                              # Required, should contain at least 1 list element
+          {
+            'action' : {                                            # Required
+              'type': 'string',                                     # Required. Can be "BOUNCE", "DROP", "DEFAULT", BYPASS_SPAM_CHECK, or MOVE_TO_JUNK
+              'parameters': { <various> }                           # Optional. For bounce, <various> can be {"bounceMessage": "message that goes in bounce mail"}
+            },
+            'recipients': list of strings,                          # Optional. Indicates list of recipients for which this action applies
+            'default': boolean                                      # Optional. Indicates whether this action applies to all recipients
+          }
+        ]}
+
+    """
+    logger.info(email_summary)
+    client = boto3.client('stepfunctions')
+    start_execution_response = client.start_execution(
```

Comment:
Awesome. Thank you for this suggestion. I knew I had to improve this logic, and your idea is something I didn't think of. I will incorporate when I get a chance.

Reply:
Looks like we need to store the executionArn to reference it in a later Lambda invocation

Reply:
There is a potential enhancement to store the executionArn in DDB (or maybe in the message in a bucket?), as the current solution has to iterate through all of the executions to find it, which may not scale to large volumes of mail. What do you think?

Reply:
Right, we need to persist the `executionArn`. Here are possible solutions:

1. Put it in DDB, for this we'll need to create a dedicated table which would add additional costs. We would also have to configure a TTL on the items so that table cleans up by itself.
2. Put it in S3 bucket with the `invcationId` as the key. S3 provides strong read-after-write consistency to as long as our `executionArn` remains same(which seems to be the case), we're good. We'll need a TTL on the bucket as well. 
3. Add it in the email itself, so you can add a new header in the email that stores the `executionArn` as it's value using `PutRawMessageContent` API. Upon the next retry read the email via `GetRawMessageContent` API and check for the header.

I'm inclined to go for option 2 because it's more cheaper, scalable and most of all simple.

Reply:
For option 2, The limitation is PUT requests 3500 per prefix, right? For very high volume scenarios objects would need to be sharded into many prefixes, I think? Would option 1 be a better option for very high volume scenarios? Or ElastiCache?

I was initially leaning towards option 3 since adding information to the headers may be useful for other use cases, but fetching/putting the entire message just to add a header seems inefficient. I don't suppose there is an option to more easily add a header?

Reply:
Quoting S3 *You can send 3,500 PUT/COPY/POST/DELETE or 5,500 GET/HEAD requests per second per prefix in an Amazon S3 bucket. There are no limits to the number of prefixes* so solution 2 can be scaled infinitely by just sharding into multiple prefixes.
 
But now that I think more, I think Option 1 would be better operationally as in if in case a customer needs to manually find a executionArn for invocationId it would be easier to search as compared to Option 2.

I will let you choose between the either of the options, both seems to have pros and cons over each other.

Reply:
To answer your question "I don't suppose there is an option to more easily add a header?", no at the moment we do not.

Reply:
I just pushed a change that incorporates DynamoDB